src/Mesh/Topologies.jl

```diff
@@ -1010,48 +1010,48 @@ end
 
 
 """    
-    grid_stretching_1d(coord_min, coord_max, Ne, stretching_type)
+    grid1d(a, b[, stretch::AbstractGridStretching]; elemsize, nelem)
```

Comment:
I'm still not 100% sold on the name `grid1d`, since we used "Grid" to mean other things.

@lcw @jkozdon any suggestions?

Reply:
I don't have a suggestion on the name but are people aware of the `meshwarp` argument to the constructor of `DiscontinuousSpectralElementGrid`?  That will change the element locations and the grid inside the element.  Here https://github.com/climate-machine/CLIMA/blob/61f69d451aa2fa3f7e905a232794ad74c5fcf7e4/examples/DGmethods/held_suarez_forcing.jl#L756-L795 is an example of where we used it in the past.

Reply:
Ah, interesting. I do hope to revamp the warp interface at some point, this is more of a stopgap until then.

Reply:
`gridrange`?

Reply:
`stretchrange`? Or can we just go with `range` with some extra arguments?

Reply:
it may not return a range (if it has been stretched)

Reply:
Make sense.

Would the more Julian way be a new range type with an [irregular step](https://github.com/JuliaLang/julia/blob/c44644442949238da70670b547312ca1ae9a9c7d/base/traits.jl#L25-L53)? (Just wondering, I'm personally fine with the array...)

Reply:
Perhaps? But these aren't performance or space critical, so probably not worth the effort.

Reply:
How about `gridspacing1d` for the name

Reply:
> it may not return a range (if it has been stretched)

is this because `range` in julia means something very specific? 
The stretching only moves the nodes. It does not change the topological structure of the grid.

Reply:
> I'm still not 100% sold on the name `grid1d`, since we used "Grid" to mean other things.
> 
> @lcw @jkozdon any suggestions?

Maybe `mesh_stretching` ?

Reply:
what about vertical_stretching.

Reply:
`vertical` wouldn't work because these are general stretching functions that we also use horizontally.

Reply:
Not sure I understand what the use case is for stretching horizontally. In the vertical it is clear.

Reply:
I guess I think of the horizontal (2D) and vertical (1D) grid decomposition in this fashion so in the horizontal we have more flexibility (in theory) to construct specific grids (via some refinement process although this doesn't exist in the code at the moment). This is why I suggested the name vertical stretching since it is only a 1D operation.  Not tied to this name but is what, say, NRL calls it in Neptune for example.

Reply:
Horizontal stretching is the closest we have to grid refinement without a grid refinement algorithm. We often use it for problems such as the squall line where the cloud is really happening within a small central portion of the domain. 

Any naming is good, but the reasoning behind having a 1D function was to use the same function whenever needed rather than building two based on the direction.

Reply:
I would agree with @smarras79 that there is nothing inherently "vertical" about a grid stretching. Our grid generator allows you to stretch in any direction.

One question is should these stretchings really be done inside the element as well. Doing the stretching here you are only stretching the element boundaries, but not the degrees of freedom inside the element.

If you want the later functionality, then this should really be a meshwarping applied elsewhere.

Reply:
> One question is should these stretchings really be done inside the element as well. Doing the stretching here you are only stretching the element boundaries, but not the degrees of freedom inside the element.
> 
> If you want the later functionality, then this should really be a meshwarping applied elsewhere.

I don't think we will ever want to change the grid spacing within an element unless there is a very specific numerical requirement for that.

Reply:
This is what @lcw was getting at with the ExponentialVerticalWarp function. This allows for the element (and the internal DOFs) to be stretched as required to maintain the theoretical accuracy.  We introduced this back in the summer for the Held-Suarez test case to better represent the boundary layer.

Reply:
I would eventually like to unify the stretching/warping into one place: perhaps instead of specifying a range in the topology constructors, it should simply be the number of elements in each dimension?

However that would be a much larger change, and I don't really want to spend too much time on it now.